mlfinlab/data_structures/imbalance_data_structures.py

```diff
@@ -0,0 +1,242 @@
+            cum_ticks += 1
+            date_time = row[0]
+            price = np.float(row[1])
+            volume = row[2]
+            cum_volume += volume
+
+            # Update high low prices
+            high_price, low_price = self._update_high_low(
+                high_price, low_price, price)
+
+            # Imbalance calculations
+            signed_tick = self._apply_tick_rule(price)
+            imbalance = self._get_imbalance(price, signed_tick, volume)
+            self.imbalance_array.append(imbalance)
+            cum_theta += imbalance
+
+            if not list_bars and np.isnan(self.expected_imbalance):
+                self.expected_imbalance = self._get_expected_imbalance(
+                    self.exp_num_ticks, self.imbalance_array)
+
+            # Update cache
+            self._update_cache(date_time, price, low_price,
+                               high_price, cum_ticks, cum_volume, cum_theta)
+
+            # Check expression for possible bar generation
+            if np.abs(cum_theta) > self.exp_num_ticks * np.abs(self.expected_imbalance):
+                self._create_bars(date_time, price,
+                                  high_price, low_price, list_bars)
+
+                self.num_ticks_bar.append(cum_ticks)
+                # Expected number of ticks based on formed bars
+                self.exp_num_ticks = ewma(np.array(
+                    self.num_ticks_bar[-self.num_prev_bars:], dtype=float), self.num_prev_bars)[-1]
+
+                self.expected_imbalance = self._get_expected_imbalance(
+                    self.exp_num_ticks * self.num_prev_bars, self.imbalance_array)
+                # Reset counters
+                cum_ticks, cum_volume, cum_theta = 0, 0, 0
+                high_price, low_price = -np.inf, np.inf
+                self.cache = []
+
+                # Update cache after bar generation (exp_num_ticks was changed after bar generation)
+                self._update_cache(date_time, price, low_price,
+                                   high_price, cum_ticks, cum_volume, cum_theta)
+        return list_bars
+
+    def _update_counters(self):
+        """
+        Updates the counters by resetting them or making use of the cache to update them based on a previous batch.
+
+        :return: Updated cum_ticks, cum_volume, cum_theta, high_price, low_price, exp_num_ticks.
+        """
+        # Check flag
+        if self.flag and self.cache:
+            latest_entry = self.cache[-1]
+
+            # Update variables based on cache
+            cum_ticks = int(latest_entry.cum_ticks)
+            cum_volume = int(latest_entry.cum_volume)
+            low_price = np.float(latest_entry.low)
+            high_price = np.float(latest_entry.high)
+            # cumulative imbalance for a particular imbalance calculation (theta_t in Prado book)
+            cum_theta = np.float(latest_entry.cum_theta)
+            # expected number of ticks extracted from prev bars
```

Comment:
What does this comment refer to? Should it still be here?

Reply:
You are right, this comment should be deleted